src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+///     : L_CURLY ws* field (ws* COMMA ws* field)* ws* (COMMA ws*)? R_CURLY
+///     | L_CURLY ws* R_CURLY
+///     ;
+fn take_struct(table: Table) -> impl Fn(&str) -> IonResult<&str, ion::Struct> {
+    move |i: &str| {
+        map(
+            preceded(
+                terminated(char(L_CURLY), eat_ws),
+                cut(alt((
+                    value(vec![], char(R_CURLY)),
+                    terminated(
+                        verify(
+                            separated_list(
+                                pair(char(','), eat_ws),
+                                terminated(take_field(table.clone()), eat_ws),
+                            ),
+                            |list: &[_]| !list.is_empty(),
+                        ),
+                        preceded(opt(pair(char(','), eat_ws)), char(R_CURLY)),
+                    ),
+                ))),
+            ),
+            |fields| ion::Struct { fields },
+        )(i)
+    }
+}
+
+/// field
+///     : field_name ws* COLON ws* annotation* entity
+///     ;
+fn take_field(table: Table) -> impl Fn(&str) -> IonResult<&str, (SymbolToken, ion::Value)> {
+    move |i: &str| {
+        map(
+            separated_pair(
+                take_field_name(table.clone()),
+                tuple((eat_ws, char(COLON), eat_ws)),
+                pair(
+                    many0(map(take_annotation(table.clone()), Some)),
+                    take_entity(table.clone()),
+                ),
+            ),
+            |(field, (annotations, value))| {
+                let annotations = if annotations.is_empty() {
+                    None
+                } else {
+                    Some(annotations)
+                };
+                (field, ion::Value { value, annotations })
+            },
+        )(i)
+    }
+}
+
+/// any_null
+///     : NULL
+///     | typed_null
+///     ;
+fn take_any_null(i: &str) -> IonResult<&str, ion::Data> {
+    alt((take_typed_null, value(ion::Data::Null, take_null)))(i)
+}
+
+/// typed_null
+///     : NULL DOT NULL
+///     | NULL DOT TYPE
+///     ;
+fn take_typed_null(i: &str) -> IonResult<&str, ion::Data> {
+    let (i, null_type) = preceded(take_null, preceded(char(DOT), alt((take_type, take_null))))(i)?;
+
+    let data = {
+        match null_type {
+            "null" => ion::Data::Null,
+            "bool" => ion::Data::Bool(None),
+            "int" => ion::Data::Int(None),
+            "float" => ion::Data::Float(None),
+            "decimal" => ion::Data::Decimal(None),
+            "timestamp" => ion::Data::Timestamp(None),
+            "string" => ion::Data::String(None),
+            "symbol" => ion::Data::Symbol(None),
+            "blob" => ion::Data::Blob(None),
+            "clob" => ion::Data::Clob(None),
+            "struct" => ion::Data::Struct(None),
+            "list" => ion::Data::List(None),
+            "sexp" => ion::Data::Sexp(None),
+            _ => unreachable!("condition covers all possible parsed results"),
```

Comment:
I made the change. The enum has to have an `as_str()` method so that the string can be extracted due to complications in sexp parsing, but the typed_null function is safer/doesn't have the `unreachable!()`